src/lua/crypto_merkle.lua

```diff
@@ -0,0 +1,363 @@
+function MT.create_merkle_tree(data_table, hashtype)
+    local N = #data_table
+    local tree = {}
+    -- hashing the data input a filling the end of the tree
+    for i = N , 2*N-1 do
+        tree[i] = _hash(data_table[ i+1-N ], hashtype)
+    end
+    -- filling the vector tree: the node in position i has as leafs the nodes in position 2i and 2i+1
+    for i = N - 1, 1, -1 do
+        local concatenated = tree[ 2*i ] .. tree[ 2*i+1 ]
+        tree[i] = _hash(concatenated, hashtype)
+    end
+    return tree
+end
+
+-- The following function is just used for testing test vectors from Frigo's RFC already hashed
+-- Here, data_table are the hash of data from the previous functions.
+-- In this case, data_table_hashed are already leaves just to insert directley into the tree.
+function MT.create_merkle_tree_from_table_of_hashes(data_table_hashed, hashtype)
+    local N = #data_table_hashed
+    local tree = {}
+
+    --in Frigo's RFC the base leaves are already hashed
+    for i = N, 2*N-1 do
+        tree[i] = data_table_hashed[ i+1-N ]
+    end
+    for i = N-1, 1, -1 do
+        local concatenated = tree[ 2*i ] .. tree[ 2*i+1 ]
+        tree[i] = _hash(concatenated, hashtype)
+    end
+
+    return tree
+end
+
+
+
+function MT.generate_proof(tree, pos)
+    local indeces_set = {}
+    local N = (#tree+1)/2
+    
+    -- checking positions of the leaves
+    if pos > N then
+        error("invalid position for leaf",2)
+    end
+    
+    pos = pos + N - 1
+
+    table.insert(indeces_set, pos)
+
+    while pos > 1 do       
+        if pos%2 == 0 then
+            table.insert(indeces_set, pos + 1 )
+        else
+            table.insert(indeces_set, pos - 1 )
+        end
+        pos = math.floor(pos/2)
+    end
+
+    local proof = {}
+    local m = #indeces_set
+
+    table.sort(indeces_set)
+
+    for i = 1, m do
+        table.insert(proof, tree[indeces_set[i]])
+    end
+
+    return proof
+end
+
+
+-- proof is a set of data already hashed since it comes from the tree in the previous function 
+-- where all data of the tree are already hashed
+-- pos is the position of the leaf to prove in the data_table (not to be considered in the tree).
+function MT.verify_proof(proof, pos, root, n , hashtype)
+    if n == 1 then
+        return proof[1]
+    end
+    
+    -- checking positions of the leaves
+    if pos > n then
+        error("invalid position for leaf",2)
+    end
+    
+    pos = pos + n - 1
+    local indeces_set = {}
+    table.insert(indeces_set, pos)
+
+    while pos > 1 do        
+        if pos%2 == 0 then
+            table.insert(indeces_set, pos + 1 )
+        else
+            table.insert(indeces_set, pos - 1 )
+        end
+        pos = math.floor(pos/2)
+    end
+
+    local set = {}
+    local m = #indeces_set
+
+    table.sort(indeces_set)
+
+    for i = 1, m do
+        table.insert(set, {proof[i], indeces_set[i]} )
+    end
+
+    local n = #proof    
+    local t = _hash(set[n-1][1] .. set[n][1]) 
```

Comment:
hashtype should always be passed to `_hash` function as second argument (here and all in the other functions call)